# Instacart Analysis: Data Engineering
by Raul Maldonado

![Data Engineering gif](https://cdn-images-1.medium.com/max/1600/1*M-yL4Huuef6gw_blv9c-Xg.gif)

## Intialization

In [64]:
import os
os.system('sh UnzipFiles.sh')

256

In [31]:
import pandas as pd
import numpy as np
import os
path = os.path.join('..','Resources','Data','RawData')

In [73]:
! ls ../Resources/Data/RawData

__MACOSX                  order_products__prior.csv products.csv
aisles.csv                order_products__train.csv sample_submission.csv
departments.csv           orders.csv


In [80]:
aislesData = pd.read_csv(f'{path}/aisles.csv')
departmentsData = pd.read_csv(f'{path}/departments.csv')
productsData = pd.read_csv(f'{path}/products.csv')
ordersData = pd.read_csv(f'{path}/orders.csv')

## Explortation

## Data Overview

1. Aisles Data
    
    a. Reference Table of Categorized Aisles
    
2. Departments Data

    a. Reference Table of Categorized Departments

3. Products Data

    a. Reference Table of Product Names with associated Aisles and Departments for every unique product
    
4. Orders Data

    a. Main Transaction Table

In [88]:
list(aislesData.columns)

['aisle_id', 'aisle']

In [79]:
departmentsData.columns

Index(['department_id', 'department'], dtype='object')

In [78]:
productsData.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id'], dtype='object')

In [81]:
ordersData.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [89]:
productsData.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13

In [108]:
ordersData.head(5)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [95]:
ordersData.describe()[:3]

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00

In [99]:
ordersData.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

We observe that column "days_since_prior_order" has a significant count of Null values, particularly 206209 null values.

In [109]:
ordersData[ordersData['user_id']==1][:5]

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

From some basic inspection, we identify if a value in the "days_since_prior_order" is null, then they are a first time request/order from a user.

## Transformation

We need to identify the following and ensure the quality of data is great for ingestion of analysis:

1. Are there any outliers?

2. Is there null data?

3. Are there irregular patterns for particular features?

4. Hows is this data digested?

5. What's the occurence of updates to data, and can system handle digestion?

6. How would we handle future schema changes?

7. How would we handle introduction of issues from data?
